lints/etsi/lint_qcstatem_psd2_policy_mandatory.go

```diff
@@ -0,0 +1,80 @@
+
+	if util.HasCertPolicy(c, policy) && !util.HasCertEtsiQcType(c, qcType) {
+		return "EU Qualified Certificate has policy " + policyStr + " but doesn't have the corresponding QcType '" + typeStr + "' (ETSI EN 319 412-4: \"Policy identifiers included in the certificate policies extension of EU Qualified Certificates shall be consistent with the EU Qualified Certificate Statements\")"
+	}
+	_, isQscdStmtPresent := util.IsQcStatemPresent(c, &util.IdEtsiQcsQcSSCD)
+	if util.HasCertPolicy(c, policy) && mustHaveQcStmtFour && !isQscdStmtPresent {
+		return "EU Qualified Certificate has policy indicating private key storage in QSCD, but the corresponding QC Statement 4 is missing (ETSI EN 319 412-4: \"Policy identifiers included in the certificate policies extension of EU Qualified Certificates shall be consistent with the EU Qualified Certificate Statements\")."
+	}
+	if util.HasCertPolicy(c, policy) && !mustHaveQcStmtFour && isQscdStmtPresent {
+		return "EU Qualified Certificate does not feature the policy indicating private key storage in QSCD, but the corresponding QC Statement 4, indicating this, is present (ETSI EN 319 412-4: \"Policy identifiers included in the certificate policies extension of EU Qualified Certificates shall be consistent with the EU Qualified Certificate Statements\")."
+	}
+	return ""
+}
+
+func (l *qcStatemPsd2PolicyMandatory) Execute(c *x509.Certificate) *lint.LintResult {
+	errStr := policyAndTypeAreConsistent(c, util.IdEtsiPolicyQcpWeb, "QCP-w", util.IdEtsiQcsQctWeb, "Web", false)
+	util.AppendToStringSemicolonDelim(&errStr, policyAndTypeAreConsistent(c, util.IdEtsiPolicyQcpNatural, "QCP-n", util.IdEtsiQcsQctEsign, "eSign", false))
+	util.AppendToStringSemicolonDelim(&errStr, policyAndTypeAreConsistent(c, util.IdEtsiPolicyQcpNaturalQscd, "QCP-n-qscd", util.IdEtsiQcsQctEsign, "eSign", true))
+	util.AppendToStringSemicolonDelim(&errStr, policyAndTypeAreConsistent(c, util.IdEtsiPolicyQcpLegal, "QCP-l", util.IdEtsiQcsQctEseal, "eSeal", false))
+	util.AppendToStringSemicolonDelim(&errStr, policyAndTypeAreConsistent(c, util.IdEtsiPolicyQcpLegalQscd, "QCP-l-qscd", util.IdEtsiQcsQctEseal, "eSeal", true))
+	if errStr != "" {
+		return &lint.LintResult{Status: lint.Error, Details: errStr}
+	}
+	_, psd2Present := util.IsQcStatemPresent(c, &util.IdEtsiPsd2Statem)
+	_, QcTypePresent := util.IsQcStatemPresent(c, &util.IdEtsiQcsQcType)
+	if psd2Present && !QcTypePresent {
+		return &lint.LintResult{Status: lint.Error, Details: "EU Qualified Certificate has PSD2 QcStatement but not the mandatory QcType Statement (ETSI EN 319 412-5: \"When the certificate is issued in accordance with Annex III or Annex IV of Regulation (EU) No 910/2014 [i.8], this statement shall be present\")."}
+	}
+	if util.HasCertAnyEtsiQcStatement(c) && len(c.PolicyIdentifiers) == 0 {
+		return &lint.LintResult{Status: lint.Error, Details: "EU Qualified Certificate must at least contain one policy identifier (ETSI EN 319 411-2: GEN-6.6.1-05)"}
```

Comment:
I don't believe this is correct, because GEN-6.6.1-05 allows for "and/or" the policy identifiers from clause 5.3 or a TSP-specific OID.

So a certificate that includes a TSP-specific OID and a QCStatement, which is valid and permitted, would trigger an error here. So I don't think this is correct.

Reply:
This checks if any (len(c.PolicyIdentifiers) == 0) policy identifier is set which I believe is correct (at least one policy identifier must be set). However we will modify this implementation to:

if len(c.PolicyIdentifiers) == 0 {
        return &lint.LintResult{Status: lint.Error, Details: "EU Qualified Certificate must at least contain one policy identifier (ETSI EN 319 411-2: GEN-6.6.1-05)"}
}

Ok? Also when this lint is triggered needs to be changed to:

func (l *qcStatemPsd2PolicyMandatory) CheckApplies(c *x509.Certificate) bool {
    return util.IsQcStatemPresent(c, &util.IdEtsiQcsQcCompliance)
}

this seems more correct. I will refactor this lint and merge main. Would it be possible to recheck?

Reply:
But the existence of an ETSI QcStatement doesn't make a certificate a Qualified certificate. Many of the QCStatements can be used for non-qualified certificates, and don't require a policy OID to be set.

So I don't think your `len(c.PolicyIdentifiers) == 0` lint would be correct here, because that'd be fine.

I think we probably want to discuss a bit more as far as what particular thing we'd like this lint to check, then work through the LintApplies. Figuring out whether or not a certificate is intended to be qualified is... obnoxiously messy (as the earlier comment captures), so it seems we should sort that out.